MobileWallet/Common/NotificationManager.swift

```diff
@@ -95,4 +155,220 @@ class NotificationManager {
+            return TariLogger.error("Failed to get wallet hex pub key", error: hexError)
+        }
+
+        guard let signature = try? wallet.signMessage("\(pubKeyHex)\(apnsDeviceToken)") else {
+            return TariLogger.error("Failed to sign message")
+        }
+
+        let url = URL(string: "\(TariSettings.shared.pushNotificationServer)/register/\(pubKeyHex)")!
+
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Accept")
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+
+        do {
+            request.httpBody = try JSONEncoder().encode(
+                TokenRegistrationServerRequest(
+                    token: apnsDeviceToken,
+                    signature: signature.hex,
+                    public_nonce: signature.nonce
+                )
+            )
+        } catch {
+            return TariLogger.error("Failed to JSON encode http body", error: error)
+        }
+
+        let onRequestError = {(requestError: Error) -> Void in
+            NotificationManager.isRegisterRequestInProgress = false
+            return TariLogger.error("Push notification register request failed", error: requestError)
+        }
+
+        NotificationManager.isRegisterRequestInProgress = true
+
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+            guard error == nil else {
+                onRequestError(error!)
+                return
+            }
+
+            guard let data = data, let response = response as? HTTPURLResponse else {
+                onRequestError(TokenServerError.unknown)
+                return
+            }
+
+            guard response.statusCode != 403 else {
+                onRequestError(TokenServerError.invalidSignature)
+                return
+            }
+
+            var responseDict: [String: Any]?
+            if let responseString = String(data: data, encoding: .utf8) {
+                if let data = responseString.data(using: .utf8) {
+                    do {
+                        responseDict = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
+                    } catch {
+                        onRequestError(error)
+                        return
+                    }
+                }
+            }
+
+            guard (200 ... 299) ~= response.statusCode else {
+                var message: String?
+                if let res = responseDict {
+                    message = res["error"] as? String
+                }
+
+                onRequestError(TokenServerError.server(response.statusCode, message: message))
+                return
+            }
+
+            guard let res = responseDict else {
+                onRequestError(TokenServerError.missingResponse)
+                return
+            }
+
+            guard let registered = res["registered"] as? Bool else {
```

Comment:
`guard let registered = responseDict?["registered"] as? Bool else {` => would make the check above redundant.

Reply:
I guess so, this way just returns a more specific error. Saying we have a json response but we didn't recognise the format. Probably overkill though, one error can catch both cases fine